src/battle_util.c

```diff
@@ -2501,16 +2501,25 @@ u8 DoFieldEndTurnEffects(void)
             while (gBattleStruct->turnSideTracker < 2)
             {
                 side = gBattleStruct->turnSideTracker;
-                gBattlerAttacker = gSideTimers[side].rainbowBattlerId;
-                if (gSideTimers[side].rainbowTimer && --gSideTimers[side].rainbowTimer == 0)
+                if (gSideStatuses[side] & SIDE_STATUS_RAINBOW)
                 {
-                    BattleScriptExecute(BattleScript_TheRainbowDisappeared);
-                    effect++;
-                }
-                else if (gSideTimers[side].rainbowTimer)
-                {
-                    BattleScriptExecute(BattleScript_RainbowContinues);
-                    effect++;
+                    for (gBattlerAttacker = 0; gBattlerAttacker < gBattlersCount; gBattlerAttacker++)
+                    {
+                        if (GetBattlerSide(gBattlerAttacker) == side)
+                            break;
+                    }
+
+                    if (--gSideTimers[side].rainbowTimer == 0)
```

Comment:
You're missing a check for `gSideTimers[side].rainbowTimer > 0`. Otherwise if the status flag is set and the timer is 0 it'll cause underflow. Same for the other pledge statuses.

Reply:
How can an underflow happen if the status flag is removed once the timer is at 0?
```
if (--gSideTimers[side].rainbowTimer == 0)
{
    gSideStatuses[side] &= ~SIDE_STATUS_RAINBOW;
    BattleScriptExecute(BattleScript_TheRainbowDisappeared);
    effect++;
}
```

Reply:
Say the user modifies the repo to sets the status flag at the start of the battle (eg. for a field effect). The current code would decrease the timer from 0 to `0xFF` and after 255 turns remove the field effect.

Reply:
I see, this makes a lot of sense.